Allow Blocks to be accessed via a long key

Spottedleaf opened the pull request:
The key can be retrieved via methods Location#toBlockKey() and
Block#getBlockKey()

World provides lookup for blocks by long key via method World#getBlockAtKey(long)

The formatting for the key is as follows:

10 bit y|27 bit z|27 bit x

The y value is considered unsigned while z and x are considered two's complement

Y range: [0, 1023]
X, Z range: [-67 108 864, 67 108 863]


Checked encoding and decoding via https://gist.github.com/Spottedleaf/74f4e241012ca2fa67d8f1c7e8e34722

DenWav commented on Spigot-API-Patches/0128-Allow-Blocks-to-be-accessed-via-a-long-key.patch:
```diff
@@ -0,0 +1,116 @@
++     * @see Block#getBlockKey()
++     */
++    public long toBlockKey() {
++        return ((long)getBlockX() & 0x3FFFFFFL) | (((long)getBlockY() & 0xFFL) << 52) | (((long)getBlockZ() & 0x3FFFFFFL) << 26);
```
I hate that you ordered it this way. I think having the two big ones first makes more sense.

aikar reviewed (changes requested):
Id rather go ahead and give X/Z the extra 2 bits incase mojang ever increases max size... be future proof.
28/28/8

And can you please update your git client and rebuild the patches.

Spottedleaf committed:
Allow Blocks to be accessed via a long key

The key can be retrieved via methods Location#toBlockKey() and
Block#getBlockKey()

World provides lookup for blocks by long key via method World#getBlockAtKey(long)

The formatting for the key is as follows:

10 bit y|27 bit z|27 bit x

The y value is considered unsigned while z and x are considered two's complement

Y range: [0, 1023]
X, Z range: [-67 108 864, 67 108 863]
